File: es_engine/regime.py
# -*- coding: utf-8 -*-
"""波动状态分类：CHOPPING / COMPRESSION / NORMAL / EXPANSION（文档 10）。纯阈值规则。"""

from __future__ import annotations
from typing import Any


def classify_regime(net_gex: float | None,
                    price: float | None,
                    pivot: float | None,
                    buffer: float,
                    call_wall: float | None,
                    put_wall: float | None,
                    em: float | None = None,
                    realized_range: float | None = None,
                    vix1d: float | None = None) -> dict[str, Any]:
    """分层判定：始终有 GEX 分量；EM/realized、VIX1D 有则细化。"""
    notes: list[str] = []
    prob: dict[str, float] = {"range_mean_revert": 0.5, "trend_expansion": 0.5}

    # 1) GEX 主导（始终可用）
    between_walls = (call_wall is not None and put_wall is not None
                     and put_wall < price < call_wall) if price is not None else False
    if net_gex is not None and net_gex > 0 and between_walls:
        prob["range_mean_revert"] = 0.65
        prob["trend_expansion"] = 0.35
        notes.append("净GEX>0 且价格在墙间 → 偏均值回归")
    if (net_gex is not None and net_gex < 0) or (price is not None and pivot is not None and price < pivot):
        prob["trend_expansion"] = 0.6
        prob["range_mean_revert"] = 0.4
        notes.append("净GEX<0 或价格在 Flip 下方 → 偏波动扩张")

    # 2) realized / EM（有 em 才启用）
    regime = "NORMAL"
    if em and realized_range is not None:
        if pivot is not None and price is not None and abs(price - pivot) <= buffer:
            regime = "CHOPPING"; notes.append("价格贴 pivot → CHOPPING")
        elif realized_range < 0.60 * em:
            regime = "COMPRESSION"; notes.append(f"实际波幅 {realized_range:.1f} < 0.6×EM({em:.1f}) → COMPRESSION")
        elif realized_range > 1.20 * em:
            regime = "EXPANSION"; notes.append(f"实际波幅 {realized_range:.1f} > 1.2×EM({em:.1f}) → EXPANSION")
    else:
        # 无 EM：用 GEX 符号给粗 regime
        if net_gex is not None and net_gex < 0:
            regime = "EXPANSION"
        elif net_gex is not None and net_gex > 0:
            regime = "CHOPPING"
        notes.append("无 EM/realized → 仅用 GEX 粗判 regime")

    # 3) VIX1D 极低位规则（文档 5.5 方法C）：vix1d<10 上调 EXPANSION 先验
    if vix1d is not None and vix1d < 10:
        prob["trend_expansion"] = min(0.75, prob["trend_expansion"] + 0.1)
        notes.append(f"VIX1D={vix1d}<10 极低位 → 短端IV无压缩空间，上调扩张先验")

    return {"regime": regime,
            "prob": prob, "notes": notes,
            "inputs": {"net_gex": net_gex, "em": em, "realized_range": realized_range,
                       "vix1d": vix1d}}

File: crates/thc-engine/src/regime.rs
//! 波动状态分类：CHOPPING / COMPRESSION / NORMAL / EXPANSION（文档 §10）。
//! 1:1 移植自 regime.py（含 VIX1D<10 先验，文档 §5.5 方法C）。

use crate::EngineConfig;
use crate::types::Regime;
use std::collections::BTreeMap;

pub struct RegimeOut {
    pub regime: Regime,
    pub prob: BTreeMap<String, f64>,
    pub notes: Vec<String>,
}

#[allow(clippy::too_many_arguments)]
pub fn classify_regime(
    net_gex: Option<f64>,
    price: Option<f64>,
    pivot: Option<f64>,
    buffer: f64,
    call_wall: Option<f64>,
    put_wall: Option<f64>,
    em: Option<f64>,
    realized_range: Option<f64>,
    vix1d: Option<f64>,
    cfg: &EngineConfig,
) -> RegimeOut {
    let mut notes: Vec<String> = Vec::new();
    let mut prob = BTreeMap::from([
        ("range_mean_revert".to_string(), 0.5),
        ("trend_expansion".to_string(), 0.5),
    ]);

    // 1) GEX 主导（始终可用）
    let between_walls = match (call_wall, put_wall, price) {
        (Some(cw), Some(pw), Some(p)) => pw < p && p < cw,
        _ => false,
    };
    if net_gex.is_some_and(|g| g > 0.0) && between_walls {
        prob.insert("range_mean_revert".into(), 0.65);
        prob.insert("trend_expansion".into(), 0.35);
        notes.push("净GEX>0 且价格在墙间 → 偏均值回归".into());
    }
    if net_gex.is_some_and(|g| g < 0.0) || (price.is_some() && pivot.is_some() && price < pivot) {
        prob.insert("trend_expansion".into(), 0.6);
        prob.insert("range_mean_revert".into(), 0.4);
        notes.push("净GEX<0 或价格在 Flip 下方 → 偏波动扩张".into());
    }

    // 2) realized / EM（有 em 才启用）
    let mut regime = Regime::Normal;
    if let (Some(em), Some(rr)) = (em, realized_range) {
        if pivot.is_some() && price.is_some() && (price.unwrap() - pivot.unwrap()).abs() <= buffer {
            regime = Regime::Chopping;
            notes.push("价格贴 pivot → CHOPPING".into());
        } else if rr < 0.60 * em {
            regime = Regime::Compression;
            notes.push(format!("实际波幅 {rr:.1} < 0.6×EM({em:.1}) → COMPRESSION"));
        } else if rr > 1.20 * em {
            regime = Regime::Expansion;
            notes.push(format!("实际波幅 {rr:.1} > 1.2×EM({em:.1}) → EXPANSION"));
        }
    } else {
        // 无 EM：用 GEX 符号给粗 regime
        if net_gex.is_some_and(|g| g < 0.0) {
            regime = Regime::Expansion;
        } else if net_gex.is_some_and(|g| g > 0.0) {
            regime = Regime::Chopping;
        }
        notes.push("无 EM/realized → 仅用 GEX 粗判 regime".into());
    }

    // 3) VIX1D 极低位规则（文档 §5.5 方法C）：vix1d<阈值 上调 EXPANSION 先验
    if vix1d.is_some_and(|v| v < cfg.vix1d_low_threshold) {
        let v: f64 = prob["trend_expansion"];
        prob.insert("trend_expansion".into(), (v + 0.1).min(0.75));
        notes.push(format!(
            "VIX1D={}<{} 极低位 → 短端IV无压缩空间，上调扩张先验",
            vix1d.unwrap(),
            cfg.vix1d_low_threshold
        ));
    }

    RegimeOut {
        regime,
        prob,
        notes,
    }
}
